src/cddl-test/Main.hs

```diff
@@ -0,0 +1,120 @@
+import           Ouroboros.Storage.FS.IO (ioHasFS)
+import           Ouroboros.Storage.EpochInfo.Impl (newEpochInfo)
+import           Ouroboros.Storage.ImmutableDB.Types (ImmutableDBError(..), ValidationPolicy(..), UserError (..))
+import qualified Ouroboros.Storage.Util.ErrorHandling as EH (exceptions)
+
+-- hard coded values
+epochSize :: Word64
+epochSize = 21600
+epochSlots :: Cardano.EpochSlots
+epochSlots = Cardano.EpochSlots epochSize
+
+givenProtocolMagic :: Cardano.ProtocolMagicId
+givenProtocolMagic = error "protocolMagic"
+
+main :: IO ()
+main = do
+    args <- getArgs
+    case args of
+        [cddlCmd, spec, dbPath] -> testBlocks cddlCmd spec dbPath Nothing
+        [cddlCmd, spec, dbPath, start, end] -> testBlocks cddlCmd spec dbPath $ Just [startEpoch .. endEpoch]
+            where
+              startEpoch = EpochNo $ read start
+              endEpoch   = EpochNo $ read end
+        _ -> putStrLn  $ concat [
+                "run the tests with :\n"
+               ,"   cddl-test CDDLCMD SPECFILE DBPATH\n"
+               ,"or cddl-test CDDLCMD SPECFILE DBPATH STARTEPOCH ENDEPOCH\n"
+               ]
+
+testBlocks :: FilePath -> FilePath -> FilePath -> Maybe [EpochNo] -> IO ()
+testBlocks cddlCmd cddlSpec chainDBPath range = do
+    db <- give epochSlots $ give givenProtocolMagic (dbArgs chainDBPath >>= openDB)
+    loop db blockList
+    where
+        (isLimitedRange, blockList) = case range of
+            Nothing -> (False , [EpochNo 0 ..])
+            Just r  -> (True  , r)
+
+        loop _ [] = return ()
+        loop db (epoch:rest) = do
+            putStr $ show  epoch
+            block <- readBlock db epoch
+            case block of
+                NoData -> error $ "failed to read ebb block: " ++ show epoch
+                FutureEpoch -> if isLimitedRange
+                    then error $ "ebb block is in the future: " ++ show epoch
+                    else putStrLn "is the future."
+                Block bytes -> do
+                    validateCBOR cddlCmd cddlSpec bytes
+                    putStrLn " OK"
+                    loop db rest
+
+validateCBOR :: FilePath -> FilePath -> ByteString -> IO ()
+validateCBOR cddlCmd cddlSpec bytes = do
+    result <- readProcessWithExitCode cddlCmd [cddlSpec, "validate", "-"] bytes
+    case result of
+        (ExitFailure _, _, err) -> error $ Char8.unpack err
+        (ExitSuccess, _, _) -> return ()
+
+dbArgs :: FilePath -> IO (ImmDbArgs IO (Block ByronConfig))
+dbArgs fp = do
+  epochInfo <- newEpochInfo (const (pure $ EpochSize epochSize))
+  return $ ImmDbArgs {
+      immDecodeHash  = Byron.decodeByronHeaderHash
+    , immDecodeBlock = Byron.decodeByronBlock epochSlots
+    , immEncodeHash  = Byron.encodeByronHeaderHash
+    , immEncodeBlock = Byron.encodeByronBlock
+    , immErr         = EH.exceptions
+    , immEpochInfo   = epochInfo
+    , immValidation  = ValidateMostRecentEpoch -- ValidateAllEpochs fails !!!
```

Comment:
Interesting! Any idea why?

Reply:
I tried it again and now it works with `ValidateAllEpochs` as well !
I removed the bogus comment but left the option at `ValidateMostRecentEpoch`.